core/modulesource.go

```diff
@@ -266,6 +282,208 @@ func (src *ModuleSource) AutomaticGitignore(ctx context.Context) (*bool, error)
 	return modCfg.Codegen.AutomaticGitignore, nil
 }
 
+// LoadContext loads a directory from the module context directory.
+//
+// If the module is local, it will load the directory from the local source
+// directly from the host.
+// In that case, the path is first resolved based on the caller's host location.
+// Then if the path is absolute, it will be relative to the context directory.
+// Otherwise, it will be relative to the module root directory.
+//
+// If the module is git, it will load the directory from the git repository
+// using its context directory.
+func (src *ModuleSource) LoadContext(ctx context.Context, dag *dagql.Server, path string, ignore []string) (inst dagql.Instance[*Directory], err error) {
+	excludes := []string{}
+	includes := []string{}
+
+	for _, p := range ignore {
+		if strings.HasPrefix(p, "!") {
+			includes = append(includes, p[1:])
+		} else {
+			excludes = append(excludes, p)
+		}
+	}
+
+	switch src.Kind {
+	case ModuleSourceKindLocal:
+		bk, err := src.Query.Buildkit(ctx)
+		if err != nil {
+			return inst, fmt.Errorf("failed to get buildkit api: %w", err)
+		}
+
+		localSourceClientMetadata, err := src.Query.NonModuleParentClientMetadata(ctx)
```

Comment:
Note: I think this will do weird things if we load a module manually using `Directory.AsModule`? In that case, a module is loading another module, and so we'd be essentially escaping back to the top-level?

At the very least, we need a test for this. We can test it using `Module.Serve`, and then sending a raw-graphql query.

Not blocking - since this is an edge case, but we should definitely cover it.